File: Dsa-day162.py
class Solution:
    def bestClosingTime(self, customers: str) -> int:
        n = len(customers)
        prefixY = [0] * (n + 1)
        for i in range(n):
            prefixY[i + 1] = prefixY[i] + (1 if customers[i] == 'Y' else 0)
        
        min_penalty = float('inf')
        best_hour = 0
        total_y = prefixY[n]
        
        for j in range(n + 1):
            open_no_cust = j - prefixY[j]
            closed_with_cust = total_y - prefixY[j]
            penalty = open_no_cust + closed_with_cust
            
            if penalty < min_penalty:
                min_penalty = penalty
                best_hour = j
        
        return best_hour

File: Dsa-day162.java
class Solution {
    public int bestClosingTime(String customers) {
        int n = customers.length();
        int[] prefixY = new int[n + 1];
        for (int i = 0; i < n; i++) {
            prefixY[i + 1] = prefixY[i] + (customers.charAt(i) == 'Y' ? 1 : 0);
        }
        
        int minPenalty = Integer.MAX_VALUE;
        int bestHour = 0;
        int totalY = prefixY[n];
        
        for (int j = 0; j <= n; j++) {
            int openNoCust = j - prefixY[j];
            int closedWithCust = totalY - prefixY[j];
            int penalty = openNoCust + closedWithCust;
            
            if (penalty < minPenalty) {
                minPenalty = penalty;
                bestHour = j;
            }
        }
        return bestHour;
    }
}